baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/PreferInputStreamTransferTo.java

```diff
@@ -0,0 +1,111 @@
+import com.google.errorprone.matchers.Matchers;
+import com.google.errorprone.matchers.method.MethodMatchers;
+import com.sun.source.tree.ExpressionTree;
+import com.sun.source.tree.MethodInvocationTree;
+import com.sun.source.tree.Tree;
+import java.util.List;
+
+@AutoService(BugChecker.class)
+@BugPattern(
+        link = "https://github.com/palantir/gradle-baseline#baseline-error-prone-checks",
+        linkType = BugPattern.LinkType.CUSTOM,
+        severity = SeverityLevel.WARNING,
+        summary = "Prefer JDK `InputStream.transferTo(OutputStream)` over utility methods such as "
+                + "`com.google.common.io.ByteStreams.copy(InputStream, OutputStream)`, "
+                + "`org.apache.commons.io.IOUtils.copy(InputStream, OutputStream)`, "
+                + "`org.apache.commons.io.IOUtils.copyLong(InputStream, OutputStream)` "
+                + "see: https://github.com/palantir/gradle-baseline/issues/2615 ",
+        explanation = "Allow for optimization when underlying input stream (such as `ByteArrayInputStream`, "
+            + "`ChannelInputStream`) overrides `long transferTo(OutputStream)` to avoid extra array allocations "
+            + "and copy larger chunks at a time (e.g. allowing 16KiB chunks via "
+            + "`ApacheHttpClientBlockingChannel.ModulatingOutputStream` from #1790).\n\n"
+            + "When running on JDK 21+, this also enables 16KiB byte chunk copies via "
+            + "`InputStream.transferTo(OutputStream)` per [JDK-8299336](https://bugs.openjdk.org/browse/JDK-8299336), "
+            + "where as on JDK < 21 and when using Guava `ByteStreams.copy` 8KiB byte chunk copies are used. "
+            + "\n\nReferences:\n\n"
+            + "  * https://github.com/palantir/hadoop-crypto/pull/586\n"
+            + "  * https://bugs.openjdk.org/browse/JDK-8299336\n"
+            + "  * https://bugs.openjdk.org/browse/JDK-8067661\n"
+            + "  * https://bugs.openjdk.org/browse/JDK-8265891\n"
+            + "  * https://bugs.openjdk.org/browse/JDK-8273038\n"
+            + "  * https://bugs.openjdk.org/browse/JDK-8279283\n"
+            + "  * https://bugs.openjdk.org/browse/JDK-8296431\n")
+public final class PreferInputStreamTransferTo extends BugChecker implements BugChecker.MethodInvocationTreeMatcher {
+
+    private static final long serialVersionUID = 1L;
+
+    private static final String ERROR_MESSAGE = "Prefer InputStream.transferTo(OutputStream)";
+
+    public static final String INPUT_STREAM = "java.io.InputStream";
+    private static final Matcher<Tree> INPUT_STREAM_MATCHER = MoreMatchers.isSubtypeOf(INPUT_STREAM);
+    public static final String OUTPUT_STREAM = "java.io.OutputStream";
+    private static final Matcher<Tree> OUTPUT_STREAM_MATCHER = MoreMatchers.isSubtypeOf(OUTPUT_STREAM);
+
+    private static final Matcher<ExpressionTree> GUAVA_BYTE_STREAM_COPY_MATCHER = MethodMatchers.staticMethod()
+            .onDescendantOf("com.google.common.io.ByteStreams")
```

Comment:
Checking specific target classes may slightly reduce the cost of this check
```suggestion
            .onClass("com.google.common.io.ByteStreams")
```

Reply:
Thanks, nice! I need to dig into the error-prone APIs a bit more (may also be worth comparing to LLM output for matchers)